styx-common/src/test/java/com/spotify/styx/util/FutureUtilTest.java

```diff
@@ -51,66 +49,83 @@
-    final IOException cause = new IOException("foo");
-    final CompletableFuture<Object> future = CompletableFuture.failedFuture(cause);
-    exception.expect(ExecutionException.class);
-    exception.expectCause(is(cause));
-    future.get();
+  public void gatherIOShouldReturnValues() throws IOException {
+    var futures = List.of(completedFuture("foo"), completedFuture("bar"));
+    assertThat(FutureUtil.gatherIO(futures, NO_TIMEOUT), contains("foo", "bar"));
   }
 
   @Test
-  public void gatherIOShouldReturnValues() {
-    var timeout = CompletableFuture.runAsync(() -> {}, delayedExecutor(30, SECONDS));
-    var futures = Map.of(
-        "foo", completedFuture("foo"),
-        "bar", completedFuture("bar"));
-    var results = FutureUtil.gatherIO(futures, timeout);
-    assertThat(results, is(Map.of(
-        "foo", Try.success("foo"),
-        "bar", Try.success("bar"))));
+  public void gatherIOShouldPropagateIOException() throws IOException {
+    var cause = new IOException("foo");
+    var futures = List.of(completedFuture("foo"), failedFuture(cause));
+    exception.expect(is(cause));
+    FutureUtil.gatherIO(futures, NO_TIMEOUT);
   }
 
   @Test
-  public void gatherIOShouldPropagateException() {
-    var timeout = CompletableFuture.runAsync(() -> {}, delayedExecutor(30, SECONDS));
+  public void gatherIOShouldPropagateException() throws IOException {
     var cause = new Exception("foo");
-    var futures = Map.of(
-        "foo", completedFuture("foo"),
-        "bar", CompletableFuture.<String>failedFuture(cause));
-    var results = FutureUtil.gatherIO(futures, timeout);
-    assertThat(results.size(), is(2));
-    assertThat(results, hasEntry("foo", Try.success("foo")));
-    assertThat(results.get("bar").getCause(), instanceOf(ExecutionException.class));
-    assertThat(results.get("bar").getCause().getCause(), is(cause));
+    var futures = List.of(completedFuture("foo"), failedFuture(cause));
+    exception.expect(RuntimeException.class);
+    exception.expectCause(is(cause));
+    FutureUtil.gatherIO(futures, NO_TIMEOUT);
+  }
+
+  @Test
+  public void gatherIOShouldPropagateTimeoutAsIOException() throws IOException {
+    var futures = List.of(completedFuture("foo"), future);
+    exception.expect(IOException.class);
+    exception.expectCause(instanceOf(TimeoutException.class));
+    FutureUtil.gatherIO(futures, TIMED_OUT);
   }
 
   @Test
-  public void gatherIOShouldApplyTimeouts() {
+  public void gatherShouldPropagateInterruptedAsIOException() throws Exception {
+    var futures = List.of(completedFuture("foo"), future);
+    Thread.currentThread().interrupt();
+    try {
+      FutureUtil.gatherIO(futures, NO_TIMEOUT);
+      fail();
+    } catch (IOException e) {
+      assertThat(e.getCause(), instanceOf(InterruptedException.class));
+    }
+    assertThat(Thread.currentThread().isInterrupted(), is(true));
+    exception.expect(InterruptedException.class);
```

Comment:
This `expect()` together with the above `fail()` is a little bit subtle to understand. Maybe add some comments to help?

Reply:
Took liberty adding the comment in 6bd84fba84358d1c67d2